src/Symfony/Component/Security/Http/EventListener/UserCheckerListener.php

```diff
@@ -0,0 +1,52 @@
+<?php
+
+namespace Symfony\Component\Security\Http\EventListener;
+
+use Symfony\Component\EventDispatcher\EventSubscriberInterface;
+use Symfony\Component\Security\Core\User\UserCheckerInterface;
+use Symfony\Component\Security\Http\Authenticator\AbstractPreAuthenticatedAuthenticator;
+use Symfony\Component\Security\Http\Event\VerifyAuthenticatorCredentialsEvent;
+
+/**
+ * @author Wouter de Jong <wouter@wouterj.nl>
+ *
+ * @final
+ * @experimental in 5.1
+ */
+class UserCheckerListener implements EventSubscriberInterface
+{
+    private $userChecker;
+
+    public function __construct(UserCheckerInterface $userChecker)
+    {
+        $this->userChecker = $userChecker;
+    }
+
+    public function preCredentialsVerification(VerifyAuthenticatorCredentialsEvent $event): void
+    {
+        if (null === $event->getUser() || $event->getAuthenticator() instanceof AbstractPreAuthenticatedAuthenticator) {
+            return;
+        }
+
+        $this->userChecker->checkPreAuth($event->getUser());
+    }
+
+    public function postCredentialsVerification(VerifyAuthenticatorCredentialsEvent $event): void
+    {
+        if (null === $event->getUser() || !$event->areCredentialsValid()) {
```

Comment:
Do we need the same `AbstractPreAuthenticatedAuthenticator` check here? Well, I want to change that - but my point is: are we missing the equivalent `if` statement here that we have in `preCredentialsVerification()`?

Reply:
No, "pre authenticated" means "the user has been authenticated before entering Symfony". That's why pre-authentication user checking should be skipped, but post-authentication user checking still applies.